### PR TITLE
Support nanos

### PR DESCRIPTION
**Pull request by rportilla-databricks**

### CODE REVIEW

File: python/tempo/tsdf.py [hunk 1]
```diff
@@ -34,15 +34,38 @@ def __init__(self, df, ts_col="event_ts", partition_cols=None, sequence_col = No
...
+      """
+      self.df = self.df.withColumn("nanos", (f.when(f.col(self.ts_col).contains("."),
+                                     f.concat(f.lit("0."), f.split(f.col(self.ts_col), '\.')[1])).otherwise(0)).cast(
+          "double")).withColumn("double_ts", f.col(self.ts_col).cast("long") + f.col("nanos"))
+
+  def __validate_ts_string(self, ts_text):
+      """ Validate the format for the string"""
+      import datetime
+      try:
+          datetime.datetime.strptime(ts_text, '%Y-%m-%dT%H:%M:%S')
```
Comment (tnixon) [inline python/tempo/tsdf.py:64]: I don't believe this format will accept a timestamp with fractional seconds. When I tested this pattern against a timestamp with fractional seconds, I got a ValueError. I'm not sure that the `strptime` function can accept nanos - the docs only talk about the %f pattern directive which supports microseconds, but not nanos (it supports up to 6 digits of fractional seconds, no more). 

I think we may need a custom REGEX here.  Try this:
```
import re
ts_pattern = "^\d{4}-\d{2}-\d{2}T| \d{2}:\d{2}:\d{2}\.\d*$"
if re.match(ts_pattern, ts_text) is None:
  raise ValueError("Incorrect data format, should be YYYY-MM-DD HH:MM:SS[.nnnnnnnn]")
```

File: python/tempo/tsdf.py [hunk 2]
```diff
@@ -34,15 +34,38 @@ def __init__(self, df, ts_col="event_ts", partition_cols=None, sequence_col = No
...
   ## Helper functions
   ##
 
-  def __validated_column(self,df,colname):
+  def __add_double_ts(self):
+      """ Add a double (epoch) version of the string timestamp out to nanos
+      """
+      self.df = self.df.withColumn("nanos", (f.when(f.col(self.ts_col).contains("."),
+                                     f.concat(f.lit("0."), f.split(f.col(self.ts_col), '\.')[1])).otherwise(0)).cast(
+          "double")).withColumn("double_ts", f.col(self.ts_col).cast("long") + f.col("nanos"))
```
Comment (tnixon) [inline python/tempo/tsdf.py:58]: You cannot simply cast the string column ts_col to long. I believe the test code in the notebook had an extra step here to convert the part ahead of the decimal to a timestamp before converting it to long in the final calculation.

**Review by tnixon — CHANGES_REQUESTED** [review]
There are a couple of problems here, see the inline comments. Was this implementation tested before creating the PR? I suggest building and testing this code in notebook (we should add testcode, too).

A couple of other questions for this:
1. the ISO timestamp format allows for either period or comma when specifying fractional seconds (it's common in European systems to use a comma) - we should support this variation.
1. If we allow for string type ts_col - it could confuse people into thinking we will support other formats (ie date only) - should we support conversion to TimestampType and DateType as well as to Double?

File: python/tempo/tsdf.py (hunk 1 above)
Comment (rportilla-databricks) [reply python/tempo/tsdf.py]: Good catch, this was tested out to seconds. It has been addressed with a new update.

File: python/tempo/tsdf.py (hunk 2 above)
Comment (rportilla-databricks) [reply python/tempo/tsdf.py]: Yes, this is a bug. It's been fixed to just compute epoch time and store that as the 'long' version.

File: python/tempo/tsdf.py (hunk 1 above)
Comment (tnixon) [reply python/tempo/tsdf.py]: Looks great

Comment (tnixon) [reply python/tempo/tsdf.py]: Actually we need the 'T' in that expression - for ISO formatted timestamps. I'm adding it back in...

File: python/tempo/tsdf.py (hunk 2 above)
Comment (tnixon) [reply python/tempo/tsdf.py]: Nice, looks good.

**Review by tnixon — APPROVED** [review]
Looks good to me!

**Pull request merged**